Add Clocking / Time capture implementation

aancel opened the pull request:
Hello,

This PR should fix #15. The basic functionalities should be covered.
What is implemented:
- [x] Quickbar (Book and Query) buttons to clock-{in,out,cancel}
- [x] Clocking logic handling in OrgFormatter
  - [x] Allow to clock-in, only if a previous clocked-in entry is not already present
  - [x] Allow to clock-out, only if a previous clocked-in entry is present, and compute the spent time 
  - [x] Allow to cancel the latest clock

What remains to be fixed:
- [x] Buttons for clock-{in,out,cancel} are currently the same alarm icon
  - Maybe having a clock icon with an arrow going in for clock-in, a clock icon with with an arrow going out for clock-out and a button with the strikethrough clock for org-cancel ?
  => using the hourglass_{top, bottom, disabled} material icons
- [ ] The regexes in OrgFormatter might be moved to OrgPatterns
- [ ] Kotlin code needs some improvement, as I do not have much experience in it
- [ ] Maybe adding buttons in the note edition UI to clock-{in,out,cancel} ?

Thanks for your comments and suggestions for improvements !

debanjum commented on app/src/main/java/com/orgzly/android/util/OrgFormatter.kt:
```diff
@@ -53,6 +54,8 @@ object OrgFormatter {
 
     private val CHECKBOXES_PATTERN = Pattern.compile("""^\s*-\s+(\[[ X]])""", Pattern.MULTILINE)
 
+    private val FULL_TIMESTAMP ="([\\[<][0-9]{4,}-[0-9]{2}-[0-9]{2} +[A-Za-z\\.]* +[0-9]{1,2}:[0-9]{2}[\\]>])"
```
Maybe we can call it `INACTIVE_DATETIMERANGE`, as it's made of [inactive timestamps](https://www.gnu.org/software/emacs/manual/html_node/org/Timestamps.html) denoting a date-time range? 
```suggestion
    private val INACTIVE_DATETIMERANGE ="([\\[<][0-9]{4,}-[0-9]{2}-[0-9]{2} +[A-Za-z\\.]* +[0-9]{1,2}:[0-9]{2}[\\]>])"
```

debanjum replied:
And a patch/diff of my suggestions for `OrgFormatter.kt` just in case :slightly_smiling_face:   
```diff
diff --git a/app/src/main/java/com/orgzly/android/util/OrgFormatter.kt b/app/src/main/java/com/orgzly/android/util/OrgFormatter.kt
index badcad64..c98d99e6 100644
--- a/app/src/main/java/com/orgzly/android/util/OrgFormatter.kt
+++ b/app/src/main/java/com/orgzly/android/util/OrgFormatter.kt
@@ -54,7 +54,7 @@ object OrgFormatter {
 
     private val CHECKBOXES_PATTERN = Pattern.compile("""^\s*-\s+(\[[ X]])""", Pattern.MULTILINE)
 
-    private val FULL_TIMESTAMP ="([\\[<][0-9]{4,}-[0-9]{2}-[0-9]{2} +[A-Za-z\\.]* +[0-9]{1,2}:[0-9]{2}[\\]>])"
+    private val INACTIVE_DATETIME_RANGE ="([\\[<][0-9]{4,}-[0-9]{2}-[0-9]{2} +[A-Za-z\\.]* +[0-9]{1,2}:[0-9]{2}[\\]>])"
 
     private const val FLAGS = Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
 
@@ -395,7 +395,7 @@ object OrgFormatter {
             val start = m.start(2) // Content start
             val end = m.end()
 
-            val CLOCKED_TIMES_P = Pattern.compile("(CLOCK: *" + FULL_TIMESTAMP + ") *(-- *" + FULL_TIMESTAMP + ")?( *=> *[0-9]{1,4}:[0-9]{2})?[\\r\\n]*")
+            val CLOCKED_TIMES_P = Pattern.compile("(CLOCK: *$INACTIVE_DATETIME_RANGE) *(-- *$INACTIVE_DATETIME_RANGE)?( *=> *[0-9]{1,4}:[0-9]{2})?[\\r\\n]*")
 
             val m2 = CLOCKED_TIMES_P.matcher(content?.subSequence(start, end))
             while (m2.find()) {
@@ -456,7 +456,7 @@ object OrgFormatter {
                 // If we find an already clocked entry, we clock-out
                 if(pos.isNotEmpty()) {
                     // Find the timestamp to compute the difference
-                    val m2 = Pattern.compile(FULL_TIMESTAMP).matcher(content)
+                    val m2 = Pattern.compile(INACTIVE_DATETIME_RANGE).matcher(content)
 
                     if (m2.find()) {
                         val clock_in = OrgDateTime.parseOrNull(m2.group(1))
```

aancel replied:
Thanks for the comments ! I've updated the code according to the review.    
I actually didn't use your suggested name `INACTIVE_DATETIME_RANGE` because the associated regex only represents one timestamp and not the range (`CLOCKED_TIMES_P` represents the whole CLOCK entry with the timerange).    
I also had to update the regex to only parse inactive timestamps (The previous one parsed both active and inactive ones).    
Please close the discussion, if you are ok with the changes.

debanjum replied:
Using `INACTIVE_DATETIME` is even better!
It seems like the way to mark a discussion as resolved is by hiding the comment, so I'll do that.

aancel replied:
Thanks !
Normally, you should have a `Resolve conversation` button at the end of the thread of your comment to do so. It should allow you to mark the discussion as done.
I have it on my side, don't you ? Or maybe those must be resolved by the maintainer ? :thinking:

debanjum replied:
Nope, don't have the Resolve conversation button on my side. Maybe you or Nevin can resolve this issue?

aancel replied:
I'll resolve the conversation then !

debanjum commented on app/src/main/java/com/orgzly/android/util/OrgFormatter.kt:
```diff
@@ -382,6 +385,143 @@ object OrgFormatter {
+            val start = m.start(2) // Content start
+            val end = m.end()
+
+            val CLOCKED_TIMES_P = Pattern.compile("(CLOCK: *" + FULL_TIMESTAMP + ") *(-- *" + FULL_TIMESTAMP + ")?( *=> *[0-9]{1,4}:[0-9]{2})?[\\r\\n]*")
```
I believe you can reference variables using the `$` symbol to build the `CLOCKED_TIMES_P` regex string
```suggestion
            val CLOCKED_TIMES_P = Pattern.compile("(CLOCK: *$FULL_TIMESTAMP) *(-- *$FULL_TIMESTAMP)?( *=> *[0-9]{1,4}:[0-9]{2})?[\\r\\n]*")
```

aancel replied:
Thanks, the change has been implemented in the last commit !

debanjum commented on app/src/main/java/com/orgzly/android/ui/notes/quickbar/QuickBar.kt:
```diff
@@ -164,7 +164,10 @@ class QuickBar(val context: Context, val inBook: Boolean) {
                     Button(R.id.quick_bar_state, R.styleable.Icons_ic_flag_24dp),
-                    Button(R.id.quick_bar_done, R.styleable.Icons_ic_outline_check_circle_24dp)
+                    Button(R.id.quick_bar_done, R.styleable.Icons_ic_outline_check_circle_24dp),
+                    Button(R.id.quick_bar_clock_in, R.styleable.Icons_ic_alarm_24dp),
```
For the different clocking state buttons would an hourglass icon work?
See example icons listed below:

* Font Awesome: [Clock-In](https://fontawesome.com/icons/hourglass-start?style=solid), [Clock-Out](https://fontawesome.com/icons/hourglass-end?style=solid)
* Noun Project: [Clock-In](https://thenounproject.com/TDsouza/collection/time/?i=780522), [Clock-Out](https://thenounproject.com/TDsouza/collection/time/?i=780521)

*Note: Couldn't find an icon for clock-cancel button in my quick search. I guess a line across the hourglass should work?*

debanjum replied:
Ended up trying the hourglass icons as clocking buttons while exploring this PR's clocking feature.

Not sure what is the right (`android` or `orgzly`) way to add new icons but sharing the results of my exploration in case it helps. Feel free to ignore if it doesn't :slightly_smiling_face:

* Icon Info
  The [clock-start](https://thenounproject.com/TDsouza/collection/time/?i=780522), [clock-out](https://thenounproject.com/TDsouza/collection/time/?i=780521) icons are from the Noun Project. 
  It's `CC-BY` licensed so requires attribution in the project. 
  Not sure where in Orgzly would be the best place to do that but icon details are below:
  `Hour Glass Start, Hour Glass Complete by Trevor Dsouza from the Noun Project`

* See [use-hourglass-icons.diff](https://github.com/orgzly/orgzly-android/files/4519929/add-hourglass-icons-and-use-for-clocking.txt)
```diff
diff --git a/app/src/main/java/com/orgzly/android/ui/notes/quickbar/QuickBar.kt b/app/src/main/java/com/orgzly/android/ui/notes/quickbar/QuickBar.kt
index 4c8c4e02..4021fed7 100644
--- a/app/src/main/java/com/orgzly/android/ui/notes/quickbar/QuickBar.kt
+++ b/app/src/main/java/com/orgzly/android/ui/notes/quickbar/QuickBar.kt
@@ -165,9 +165,9 @@ class QuickBar(val context: Context, val inBook: Boolean) {
                     Button(R.id.quick_bar_deadline, R.styleable.Icons_ic_alarm_24dp),
                     Button(R.id.quick_bar_state, R.styleable.Icons_ic_flag_24dp),
                     Button(R.id.quick_bar_done, R.styleable.Icons_ic_outline_check_circle_24dp),
-                    Button(R.id.quick_bar_clock_in, R.styleable.Icons_ic_alarm_24dp),
-                    Button(R.id.quick_bar_clock_out, R.styleable.Icons_ic_alarm_24dp),
-                    Button(R.id.quick_bar_clock_cancel, R.styleable.Icons_ic_alarm_24dp)
+                    Button(R.id.quick_bar_clock_in, R.styleable.Icons_ic_hourglass_start_black),
+                    Button(R.id.quick_bar_clock_out, R.styleable.Icons_ic_hourglass_stop_black),
+                    Button(R.id.quick_bar_clock_cancel, R.styleable.Icons_ic_hourglass_cancel_black)
             )
 
             private val LEFT_IN_BOOK = listOf(
diff --git a/app/src/main/res/drawable/ic_hourglass_cancel_black.xml b/app/src/main/res/drawable/ic_hourglass_cancel_black.xml
new file mode 100644
index 00000000..25765d9c
--- /dev/null
+++ b/app/src/main/res/drawable/ic_hourglass_cancel_black.xml
@@ -0,0 +1,9 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="100"
+    android:viewportHeight="100">
+  <path
+      android:fillColor="#FF000000"
+      android:pathData="M86.125,95.688h-5.957c0.525,-3.746 1.245,-10.342 1.245,-18.436 0,-8.884 -8.747,-14.251 -15.134,-18.169 -2.271,-1.394 -5.363,-3.291 -5.745,-4.148 0.438,-0.92 3.5,-2.79 5.753,-4.166 6.384,-3.899 15.126,-9.239 15.126,-18.021 0,-8.709 -0.834,-15.694 -1.361,-19.247h6.073c1.646,0 2.979,-1.334 2.979,-2.979 0,-1.645 -1.334,-2.979 -2.979,-2.979L55,7.543 23.875,7.543c-1.646,0 -2.979,1.334 -2.979,2.979 0,1.646 1.334,2.979 2.979,2.979h6.074c-0.528,3.553 -1.362,10.538 -1.362,19.247 0,8.783 8.743,14.123 15.126,18.021 2.237,1.366 5.272,3.221 5.743,4.147 -0.46,0.93 -3.497,2.793 -5.735,4.167 -6.387,3.918 -15.134,9.285 -15.134,18.169 0,8.094 0.72,14.689 1.245,18.436h-5.957c-1.646,0 -2.979,1.334 -2.979,2.979 0,1.645 1.334,2.979 2.979,2.979L55,101.646 86.125,101.646c1.646,0 2.979,-1.334 2.979,-2.979 0,-1.645 -1.333,-2.979 -2.979,-2.979zM76.123,95.688L55,95.688 33.876,95.688C33.37,92.2 32.586,85.475 32.586,77.252c0,-6.645 7.333,-11.145 13.226,-14.76 4.357,-2.674 7.505,-4.605 7.666,-7.376C53.65,52.151 50.356,50.139 45.797,47.355 39.911,43.76 32.586,39.286 32.586,32.747 32.586,23.845 33.508,16.68 33.998,13.5L55,13.5 76.002,13.5c0.489,3.18 1.411,10.345 1.411,19.247 0,6.539 -7.325,11.013 -13.211,14.608 -4.56,2.785 -7.854,4.796 -7.683,7.761 0.162,2.771 3.311,4.702 7.668,7.376 5.893,3.615 13.226,8.115 13.226,14.76 0,8.207 -0.784,14.939 -1.29,18.436z"/>
+</vector>
diff --git a/app/src/main/res/drawable/ic_hourglass_start_black.xml b/app/src/main/res/drawable/ic_hourglass_start_black.xml
new file mode 100644
index 00000000..cce5ad96
--- /dev/null
+++ b/app/src/main/res/drawable/ic_hourglass_start_black.xml
@@ -0,0 +1,12 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="100"
+    android:viewportHeight="100">
+  <path
+      android:fillColor="#FF000000"
+      android:pathData="M86.125,95.688h-5.957c0.525,-3.746 1.245,-10.342 1.245,-18.436c0,-8.884 -8.747,-14.251 -15.134,-18.169c-2.271,-1.394 -5.363,-3.291 -5.745,-4.148c0.438,-0.92 3.5,-2.79 5.753,-4.166c6.384,-3.899 15.126,-9.239 15.126,-18.021c0,-8.709 -0.834,-15.694 -1.361,-19.247h6.073c1.646,0 2.979,-1.334 2.979,-2.979s-1.334,-2.979 -2.979,-2.979H55H23.875c-1.646,0 -2.979,1.334 -2.979,2.979c0,1.646 1.334,2.979 2.979,2.979h6.074c-0.528,3.553 -1.362,10.538 -1.362,19.247c0,8.783 8.743,14.123 15.126,18.021c2.237,1.366 5.272,3.221 5.743,4.147c-0.46,0.93 -3.497,2.793 -5.735,4.167c-6.387,3.918 -15.134,9.285 -15.134,18.169c0,8.094 0.72,14.689 1.245,18.436h-5.957c-1.646,0 -2.979,1.334 -2.979,2.979s1.334,2.979 2.979,2.979H55h31.125c1.646,0 2.979,-1.334 2.979,-2.979S87.771,95.688 86.125,95.688zM76.123,95.688H55H33.876c-0.506,-3.488 -1.29,-10.213 -1.29,-18.436c0,-6.645 7.333,-11.145 13.226,-14.76c4.357,-2.674 7.505,-4.605 7.666,-7.376c0.172,-2.965 -3.122,-4.977 -7.681,-7.761c-5.886,-3.595 -13.211,-8.069 -13.211,-14.608c0,-8.902 0.922,-16.067 1.412,-19.247H55h21.002c0.489,3.18 1.411,10.345 1.411,19.247c0,6.539 -7.325,11.013 -13.211,14.608c-4.56,2.785 -7.854,4.796 -7.683,7.761c0.162,2.771 3.311,4.702 7.668,7.376c5.893,3.615 13.226,8.115 13.226,14.76C77.413,85.459 76.629,92.191 76.123,95.688z"/>
+  <path
+      android:fillColor="#FF000000"
+      android:pathData="M63.542,46.413c5.464,-3.338 12.265,-7.491 12.265,-13.561c0,-8.264 -0.856,-14.915 -1.311,-17.867H55H35.505c-0.455,2.952 -1.311,9.603 -1.311,17.867c0,6.07 6.8,10.224 12.264,13.561c4.233,2.585 7.29,4.284 7.13,7.036h2.823C56.252,50.697 59.311,48.998 63.542,46.413z"/>
+</vector>
diff --git a/app/src/main/res/drawable/ic_hourglass_stop_black.xml b/app/src/main/res/drawable/ic_hourglass_stop_black.xml
new file mode 100644
index 00000000..88b0df8d
--- /dev/null
+++ b/app/src/main/res/drawable/ic_hourglass_stop_black.xml
@@ -0,0 +1,12 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="24dp"
+    android:height="24dp"
+    android:viewportWidth="100"
+    android:viewportHeight="100">
+  <path
+      android:fillColor="#FF000000"
+      android:pathData="M86.125,95.688h-5.957c0.525,-3.746 1.245,-10.342 1.245,-18.436c0,-8.884 -8.747,-14.251 -15.134,-18.169c-2.271,-1.394 -5.363,-3.291 -5.745,-4.148c0.438,-0.92 3.5,-2.79 5.753,-4.166c6.384,-3.899 15.126,-9.239 15.126,-18.021c0,-8.709 -0.834,-15.694 -1.361,-19.247h6.073c1.646,0 2.979,-1.334 2.979,-2.979c0,-1.645 -1.334,-2.979 -2.979,-2.979H55H23.875c-1.646,0 -2.979,1.334 -2.979,2.979c0,1.646 1.334,2.979 2.979,2.979h6.074c-0.528,3.553 -1.362,10.538 -1.362,19.247c0,8.783 8.743,14.123 15.126,18.021c2.237,1.366 5.272,3.221 5.743,4.147c-0.46,0.93 -3.497,2.793 -5.735,4.167c-6.387,3.918 -15.134,9.285 -15.134,18.169c0,8.094 0.72,14.689 1.245,18.436h-5.957c-1.646,0 -2.979,1.334 -2.979,2.979c0,1.646 1.334,2.979 2.979,2.979H55h31.125c1.646,0 2.979,-1.334 2.979,-2.979C89.104,97.021 87.771,95.688 86.125,95.688zM76.123,95.688H55H33.876c-0.506,-3.489 -1.29,-10.213 -1.29,-18.436c0,-6.645 7.333,-11.145 13.226,-14.76c4.357,-2.674 7.505,-4.605 7.666,-7.376c0.172,-2.965 -3.122,-4.977 -7.681,-7.761c-5.886,-3.595 -13.211,-8.069 -13.211,-14.608c0,-8.902 0.922,-16.067 1.412,-19.247H55h21.002c0.489,3.18 1.411,10.345 1.411,19.247c0,6.539 -7.325,11.013 -13.211,14.608c-4.56,2.785 -7.854,4.796 -7.682,7.761c0.162,2.771 3.31,4.702 7.667,7.376c5.893,3.615 13.226,8.115 13.226,14.76C77.413,85.459 76.629,92.192 76.123,95.688z"/>
+  <path
+      android:fillColor="#FF000000"
+      android:pathData="M63.281,63.756c-3.928,-2.41 -6.765,-4.314 -6.911,-6.813h-2.742c-0.144,2.498 -2.982,4.402 -6.91,6.813c-5.312,3.259 -11.921,7.315 -11.921,13.305c0,7.412 0.706,13.473 1.163,16.618H55h19.04c0.456,-3.15 1.163,-9.221 1.163,-16.618C75.203,71.071 68.593,67.015 63.281,63.756z"/>
+</vector>
diff --git a/app/src/main/res/values/attrs.xml b/app/src/main/res/values/attrs.xml
index 70e5f5e5..085d75eb 100644
--- a/app/src/main/res/values/attrs.xml
+++ b/app/src/main/res/values/attrs.xml
@@ -146,6 +146,10 @@
         <attr name="ic_warning_24dp" format="reference"/>
         <attr name="ic_widgets_24dp" format="reference"/>
 
+        <attr name="ic_hourglass_start_black" format="reference"/>
+        <attr name="ic_hourglass_stop_black" format="reference"/>
+        <attr name="ic_hourglass_cancel_black" format="reference"/>
+
         <attr name="oic_dropbox_linked" format="reference"/>
         <attr name="oic_dropbox_not_linked" format="reference"/>
         <attr name="oic_git_48dp" format="reference"/>
diff --git a/app/src/main/res/values/styles.xml b/app/src/main/res/values/styles.xml
index d9251fa9..12e600e8 100644
--- a/app/src/main/res/values/styles.xml
+++ b/app/src/main/res/values/styles.xml
@@ -168,6 +168,10 @@
         <item name="ic_warning_24dp">@drawable/ic_warning_black_24dp</item>
         <item name="ic_widgets_24dp">@drawable/ic_widgets_black_24dp</item>
 
+        <item name="ic_hourglass_start_black">@drawable/ic_hourglass_start_black</item>
+        <item name="ic_hourglass_stop_black">@drawable/ic_hourglass_stop_black</item>
+        <item name="ic_hourglass_cancel_black">@drawable/ic_hourglass_cancel_black</item>
+
         <item name="oic_dropbox_linked">@drawable/cic_dropbox_blue_48px</item>
         <item name="oic_git_48dp">@drawable/cic_git_black_48dp</item>
         <item name="oic_dropbox_not_linked">@drawable/cic_dropbox_black_48px</item>
```

aancel replied:
The hourglass icons are a great idea :+1:      
I think that we should wait for @nevenz 's input on this side to see if it fits in the general design used for the project (and notably if he used specific icon sources).

aancel replied:
Following @nevenz 's recommandation, I've update the icons in de2c09f. They should be working in light, dark and black themes.

lckarssen commented:
@aancel is there an easy way for me to try your clocking functionality? I have never created an Android package myself, so I was wondering if you have package I can try via e.g. F-droid or directly as a `.apk`.

aancel commented:
Hello @lckarssen,
Regarding feature testing, I don't think that it would be a good idea to install a custom debug build apk, if you already have Orgzly properly setup on your phone (It might probably not install or break the update to newer versions).    

If you are willing to try via Android Studio, it is quite easy to test the feature via the android emulator.    
You would need to:
- Install Android Studio
- Import the project via File -> New -> Project from version control -> Git, then type in the url of my fork
- You would then need to make sure that the correct branch is selected via VCS -> Git
- Then you can create a Virtual Machine via Tools -> AVD Manager (I usually use Pixel phones for this) or just plug in your phone
- And then run the app (Should be bound to Shift+F10)

lckarssen commented:
Thank you for the instructions, @aancel. I'll have a look at Android Studio.

aancel commented:
@lckarssen In case you are having too much trouble setting everything up, I can still send you an apk if you would like (I'd just need a way to send it to you)

debanjum commented:
@lckarssen you can use [APK](https://github.com/debanjum/orgzly-android/suites/725260416/artifacts/7222772) with @aancel's clocking feature created by the [Workflow](https://github.com/debanjum/orgzly-android/pull/2/checks?check_run_id=714096621) from [this PR](https://github.com/orgzly/orgzly-android/pull/721)

allentiak commented:
@aancel Just in case you missed it, @nevenz  did some comments on your PR.

aancel commented:
Sorry, life got a bit badly on the way those last months.    
I should be able to start working on this feature again soon (in the coming days/weeks).    
I still really want this feature too in Orgzly !

allentiak commented:
@aancel Looking forward to those changes!

aancel commented:
@nevenz I've noticed an unexpected behavior in the drawerPattern function in OrgFormatter:
```kotlin
    private fun drawerPattern(name: String) = Pattern.compile(
            """^[ \t]*:($name):[ \t]*\n(.*?)\n[ \t]*:END:[ \t]*$""",
            Pattern.CASE_INSENSITIVE or Pattern.MULTILINE or Pattern.DOTALL) 
```

If I do a clock-in, then clock-cancel, then clock-in, I end up having 2 LOGBOOK drawers:
```
:LOGBOOK:
CLOCK: [2020-10-25 Sun 17:43]
:END:

:LOGBOOK:
:END:
```

If I update the drawerPattern function to add a `*` after the second `\n`, I no longer have this behavior:
```kotlin
    private fun drawerPattern(name: String) = Pattern.compile(
            """^[ \t]*:($name):[ \t]*\n(.*?)\n*[ \t]*:END:[ \t]*$""",
            Pattern.CASE_INSENSITIVE or Pattern.MULTILINE or Pattern.DOTALL)
```

As this modification could have an impact elsewhere, I'd rather ask you whether I should push it or not

aancel commented on app/src/main/java/com/orgzly/android/ui/notes/NotesFragment.kt:
```diff
@@ -171,7 +171,7 @@ abstract class NotesFragment : Fragment(), TimestampDialogFragment.OnDateTimeSet
         fun onScheduledTimeUpdateRequest(noteIds: Set<Long>, time: OrgDateTime?)
         fun onDeadlineTimeUpdateRequest(noteIds: Set<Long>, time: OrgDateTime?)
-
+        fun onClockingUpdateRequest(noteIds: Set<Long>, type: Int?)
```
I've updated the code to use the SharedMainActivityViewModel, but I'm not very sure if the way it is implemented is correct, so if you can give me feedback on this, that would be great !     
I notably didn't know how to implement the calls to onError and onSuccess as of now.

nevenz commented:
> If I do a clock-in, then clock-cancel

Org mode removes the drawer when it ends up being empty after this.

> If I update the drawerPattern function to add a * after the second \n, I no longer have this behavior:

I think that would match:

```
:LOGBACK:
foo:END:
```

lytex commented:
When this branch is merged with master, the build fails with the following error:

```
Error while processing ..../orgzly-android/app/src/main/res/drawable/ic_hourglass_top.xml : Invalid color value ?attr/text_primary_color
```

Reverting this specific commit `7c6b75f6 - Update Gradle to v6.5 and plugin to v4.1.1` makes it work again. No idea why is this happening.

nevenz commented:
This seems to be needed, due to the use of resources in vector drawables:

```diff
diff --git a/app/build.gradle b/app/build.gradle
index b4ce864b..ee8be29c 100644
--- a/app/build.gradle
+++ b/app/build.gradle
@@ -37,6 +37,8 @@ android {
                 ]
             }
         }
+
+        vectorDrawables.useSupportLibrary = true
     }
 
     buildFeatures {
```

aancel commented:
@nevenz I've updated the code following your latest reviews.    
- I've added the change you requested in the build.gradle file
- I've also updated the clock cancel code to delete the drawer when it ends up being empty after a cancel (not so sure about the way I implemented it)

aancel commented:
Could we resolve or iterate on the discussions that have been addressed so I can see what remains to be done in this PR, please ? :thinking:

I think that there is at least this comment from @nevenz https://github.com/orgzly/orgzly-android/pull/691#pullrequestreview-433394906 that asks for several improvements.     
TBH I think that I could handle all the UI related stuff, but it would probably take me some time to handle the part with the `clock_range_id` in the `notes` table, as I haven't touched any db code yet.     
So how should we handle this ?

nevenz commented:
> I think that I could handle all the UI related stuff, but it would probably take me some time to handle the part with the clock_range_id in the notes table, as I haven't touched any db code yet.

I can do that part. I think it's important to be able to quickly jump to currently clocked note right from the start, which does require keeping DB in sync with `CLOCK` entries.

allentiak commented:
Just curious... Any news regarding this?

lytex commented:
I cannot get this build to work if I don't include this changes, I think they are missing:

```diff
diff --git a/app/src/main/res/layout/list_widget.xml b/app/src/main/res/layout/list_widget.xml
index f60425f6..4b58bb7c 100644
--- a/app/src/main/res/layout/list_widget.xml
+++ b/app/src/main/res/layout/list_widget.xml
@@ -1,10 +1,10 @@
 <?xml version="1.0" encoding="utf-8"?>
 
-<LinearLayout
-    xmlns:android="http://schemas.android.com/apk/res/android"
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
     android:layout_width="match_parent"
     android:layout_height="match_parent"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
     android:orientation="vertical">
 
     <LinearLayout
@@ -22,7 +22,7 @@
             android:layout_width="48dp"
             android:layout_height="match_parent"
             android:layout_gravity="center_vertical"
-            android:src="@drawable/ic_logo_for_widget" />
+            app:srcCompat="@drawable/ic_logo_for_widget" />
 
         <LinearLayout
             android:id="@+id/list_widget_header_bar"
```

xiaoruoruo commented:
One suggestion is to add a persistent notification when there is an active clock. Emacs have that in the mode-line to remind you. On Android it would be good to have a persistent notification to know there is an active clock going on. And that could have a button to quickly clock out.

MisterY commented:
> One suggestion is to add a persistent notification when there is an active clock. 

Sure. But that can be requested separately. Otherwise, we may not live long enough to see this feature coming to life.

allentiak commented:
@aancel @nevenz Specifically... What do you think would be needed to move this conversation forward, so this PR could eventually be merged? Whereas I don't know much Kotlin, I might be able to help...

aancel commented:
Hello,
Regarding this PR, I think that @debanjum still has a discussion opened, could you close it if the changes are ok for you ?
Regarding your comment @lytex, I've just checked and it still seems to be building on my side, with the latest android-studio, are you up to date or maybe using a different setup that could cause the update you mention to be required ? 🤔
I think that the suggestion from @xiaoruoruo to have a persistent notification is a very good idea, maybe this could be a good idea to open an issue to keep this suggestion in the backlog ? And do this as a second step !
Please @nevenz do tell me, if I have to apply some additional modifications, I'll try my best to take some time for them !

debanjum commented:
> Hello,
> Regarding this PR, I think that @debanjum still has a discussion opened, could you close it if the changes are ok for you ?

Not sure how to close the discussion but that discussion shouldn't block this PR. And the changes look good to me! 😃

aancel commented:
I've resolved the discussion, it was indeed on an old diff so not blocking, sorry for the ping 😅

ugurbolat commented:
Hi, any news about this feature?

seylerius commented:
Are there any remaining blockers on merging this feature? I'm very eager to see it released, though I'll likely build the APK for myself this week to start using it early.

michaelsjackson commented:
Can this feature be used for doing something like this https://github.com/orgzly/orgzly-android/issues/869 ?

allentiak commented:
@nevenz Any news regarding this? https://github.com/orgzly/orgzly-android/pull/691#issuecomment-796368978

stites commented:
I would really like to see this officially supported.

Since the CI artifacts can no longer be downloaded, I've merged this into master on my fork ([see commit history here](https://github.com/stites/orgzly-android/pull/1)) and the apk generated from this action can be found [on this page](https://github.com/stites/orgzly-android/actions/runs/1143227029) (different PR, but shas are identical). You can [directly downloaded the generated apks here](https://github.com/stites/orgzly-android/suites/3530995979/artifacts/84429486).

(heads up: took me a minute to realize that the quickbar is activated from swiping right, or left-right, on an item)

allentiak commented:
@nevenz Do you think this PR could be merged as-is?
BTW: Is there anyone else with the right permissions to do this?

@stites I would like to try your build. Which artifact should be installed? There are two: "fdroid" and "premium"...

stites commented:
@allentiak premium is the paid version of this app which gives you access to the dropbox backend. Paying for the premium app through the app store is the only way to donate to this project (you can also try to ping @nevenz in https://github.com/orgzly/orgzly-android/issues/710 to change this). the fdroid version doesn't include the dropbox backend.

nevenz commented:
> @nevenz it seems that everything is there, could this be merged?

I took a quick look again, there are still some UI things I need to do (too many icons in the swipe menu, no buttons in the action bar, jumping to the currently clocked task would be useful right from the start, etc.).

schlagenhauf commented:
I would like to help, if you describe what you have in mind exactly.

daniebker commented:
I've been using this branch for a little while. One minor thing. If you clock in and immediately out it will still log a 0:00 time stamp in the logbook. Normal org-mode behaviour doesn't log anything.

omidmnz commented:
> I've been using this branch for a little while. One minor thing. If you clock in and immediately out it will still log a 0:00 time stamp in the logbook. Normal org-mode behaviour doesn't log anything.

I don't think that's the default behavior in Org mode. The variable `org-clock-out-remove-zero-time-clocks` controls it.

n14s commented:
> You can [directly downloaded the generated apks here](https://github.com/stites/orgzly-android/suites/3530995979/artifacts/84429486).

The artifact expired. Can you recreate it @stites?

stites commented:
bumped pr: https://github.com/stites/orgzly-android/pull/4
actions page: https://github.com/stites/orgzly-android/actions/runs/1626843482
direct download of artifact: https://github.com/stites/orgzly-android/suites/4769085519/artifacts/132700825

UdayvirThind commented:
Is it possible to have something like:
![Screenshot_20220215-123425_Boosted](https://user-images.githubusercontent.com/74961020/154010079-2282030d-a185-4aca-b756-1b99487a3c6e.png)
This would practically replace my entire time management workflow

allentiak commented:
> I would like to help, if you describe what you have in mind exactly.

@nevenz Ping?

fkgruber commented:
Is there a apk with this feature that we can try?

fkgruber commented:
Hi @aancel 
I checked out your feat-clock-time-capture branch (https://github.com/aancel/orgzly-android/tree/feat-clocking-time-capture) and build it and run in on android studio but I don't see any additional buttons to clock in or out. Maybe I'm missing something. Where do I find the clock in/clock out buttons?
thanks
F




> If you are willing to try via Android Studio, it is quite easy to test the feature via the android emulator. You would need to:
> 
> * Install Android Studio
> * Import the project via File -> New -> Project from version control -> Git, then type in the url of my fork
> * You would then need to make sure that the correct branch is selected via VCS -> Git
> * Then you can create a Virtual Machine via Tools -> AVD Manager (I usually use Pixel phones for this) or just plug in your phone
> * And then run the app (Should be bound to Shift+F10)

lytex commented:
Hi @fkgruber

You have to swipe right on the heading you want to clock in. You should see an hourglass with the sand on the top to clock-in, a hourglass with the sand on the bottom to clock-out and a crossed out hourglass to cancel clocking

fkgruber commented:
> Hi @fkgruber
> 
> You have to swipe right on the heading you want to clock in. You should see an hourglass with the sand on the top to clock-in, a hourglass with the sand on the bottom to clock-out and a crossed out hourglass to cancel clocking

Thanks! that worked

allentiak commented:
Just curious...
@nevenz 
What else is needed so this can be merged?

leejh903 commented:
I connected my android phone and installed apk. But this version does not have Dropbox repostiory 😢 I eager to this PR to be merged after rebasing the master

nevenz commented:
@allentiak 
> What else is needed so this can be merged?

Sorry, I'll check it again after v1.8.10.
